dutch-auction/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+#![no_std]
+#![feature(const_btree_new)]
+
+use codec::Encode;
+use gstd::{msg, prelude::*, exec::block_timestamp, ActorId};
+use primitive_types::U256;
+use nft_example_io;
+
+pub mod state;
+pub use state::{State, StateReply};
+
+pub use auction_io::{Action, Event, InitConfig, CreateConfig};
+
+const ZERO_ID: ActorId = ActorId::new([0u8; 32]);
+const DURATION: u64 = 7 * 24 * 60 * 60 * 1000;
+
+#[derive(Debug)]
+pub struct NFT {
+    pub token_id: U256,
+    pub owner: ActorId,
+    pub contract_id: ActorId,
+}
+
+#[derive(Debug)]
+pub struct Auction {
+    pub nft: NFT,
+    pub starting_price: U256,
+    pub discount_rate: U256,
+    pub is_using: bool,
+    pub start_at: u64,
```

Comment:
I think `started_at` is redundant variable (since there is an `expires_at` and `duration`)

Reply:
If it's preferable in smart contracts to carry out extra calculations instead of storing variables, then it is possible, of course, but this will make the code less readable

Reply:
I vote for avoiding reduntant fields too. We can forget to calculate some of them that yields in wrong behavior.

Reply:
I've threw out the duration